Examples/Integration/IntegrationUITests/Internal/BaseIntegrationTests.swift

```diff
@@ -0,0 +1,128 @@
+    self.app.launch()
+    self.logs = self.app.staticTexts["composable-architecture.debug.logs"]
+  }
+
+  override func tearDown() {
+    super.tearDown()
+    if let (file, line) = self._expectRuntimeWarnings {
+      XCTAssert(
+        self.app.staticTexts["Runtime warning"].waitForExistence(timeout: 1),
+        "Expected runtime warning(s)",
+        file: file,
+        line: line
+      )
+    } else {
+      XCTAssertFalse(self.app.staticTexts["Runtime warning"].exists)
+    }
+  }
+
+  func clearLogs() {
+    XCUIDevice.shared.system.open(URL(string: "integration:///clear-logs")!)
+  }
+
+  func assertLogs(
+    _ expectedLogs: String,
+    file: StaticString = #file,
+    line: UInt = #line
+  ) {
+    defer { self.clearLogs() }
+    XCTAssertNoDifference(
+      self.logs.label,
+      expectedLogs,
+      file: file,
+      line: line
+    )
+  }
+
+  func assertLogs(
+    _ expectedLogs: [Log],
+    file: StaticString = #file,
+    line: UInt = #line
+  ) {
+    defer { self.clearLogs() }
+
+    print(self.logs.label)
+    var actualLogLines = self.logs.label.split(separator: "\n").map(String.init)
+    var foundFailure = false
+    for log in expectedLogs {
+      switch log {
+      case let .exact(logs):
+        let logLines = logs.split(separator: "\n").map(String.init)
+        if !actualLogLines.prefix(logLines.count).elementsEqual(logLines) {
+          foundFailure = true
+          break
+        }
+        actualLogLines.removeFirst(min(logLines.count, actualLogLines.count))
+      case let .matching(logs):
+        guard let logLines = try? logs.split(separator: "\n").map({ try Regex(String($0)) })
+        else {
+          foundFailure = true
+          break
+        }
+        if zip(actualLogLines, logLines).allSatisfy({ $0.wholeMatch(of: $1) != nil }) {
+          actualLogLines.removeFirst(min(logLines.count, actualLogLines.count))
+        }
+      case let .optional(logs):
+        let logLines = logs.split(separator: "\n").map(String.init)
+        if actualLogLines.prefix(logLines.count).starts(with: logLines) {
+          actualLogLines.removeFirst(min(logLines.count, actualLogLines.count))
+        }
+      case let .unordered(logs):
+        let logLines = logs.split(separator: "\n").map(String.init).sorted()
+        guard logLines.count <= actualLogLines.count else {
+          foundFailure = true
+          break
+        }
+        for logLine in logLines {
+          guard let index = actualLogLines.firstIndex(of: logLine)
+          else {
+            foundFailure = true
+            break
+          }
+          actualLogLines.remove(at: index)
+        }
+      }
+    }
+
+    if foundFailure || !actualLogLines.isEmpty {
+      XCTAssertNoDifference(
+        self.logs.label,
+        expectedLogs.map(\.description).joined(separator: "\n"),
+        file: file,
+        line: line
+      )
+    }
+  }
+}
+
+enum Log: ExpressibleByStringLiteral, CustomStringConvertible {
```

Comment:
I think we should ditch all of this and go with unordered assertions by default, _and_ we should use inline snapshot testing to do the assertions.